server/src/main/java/io/deephaven/server/session/SessionServiceGrpcImpl.java

```diff
@@ -177,13 +167,45 @@ public void exportFromTicket(
 
         final SessionState.ExportObject<Object> source = ticketRouter.resolve(
                 session, request.getSourceId(), "sourceId");
-        session.newExport(request.getResultId(), "resultId")
+        final SessionState.ExportObject<Object> destination = session.newExport(request.getResultId(), "resultId")
                 .require(source)
                 .onError(responseObserver)
-                .submit(() -> {
-                    GrpcUtil.safelyComplete(responseObserver, ExportResponse.getDefaultInstance());
-                    return source.get();
+                .submit(source::get);
+        session.nonExport()
+                .require(destination)
+                .onError(responseObserver)
+                .submit(() -> GrpcUtil.safelyComplete(responseObserver, ExportResponse.getDefaultInstance()));
```

Comment:
I'm not sure I follow - source.get() is strictly before the safelyComplete call, since the source.get() will be written to `destination`, and the safelyComplete won't complete until `destination` is exported? If the source.get() fails, destination will be left in an error state, so safelyComplete won't be called - we would attempt to call the responseObserver.onError(...), but the second invocation would fail, so only the first error would be sent.

In the case of the publish method below, two distinct submit()s are required, one that depends on the source and one that depends on the destination. It is somewhat non-standard to require(destination) and not respond to the client until that happens though, since we know that destination must be an export ticket, so the client can already race this call. Getting a successful result here only depends on the source being a success - if we think this pattern is required here, it is likely required in many more places as well? Not wrong, just not necessary.

Reply:
After talking with Devin, I agree that the simpler fix would be to call source.get(), store in a temp variable, and then call safelyComplete after that source.get() was successful. I don't think this extra nonExport() buys us anything - if we think it is important, we must add it in many other places. I also don't think it is doing anything wrong, and cannot introduce an error we aren't already handling safely (from double-calling onError).

Reply:
Yeah; Devin is right. This covers his original concern and the non-export bought us nothing.